doc/dev/autograph.rst

```diff
@@ -0,0 +1,653 @@
+feature of Catalyst, AutoGraph, instead allows Catalyst to work
+with **native Python control flow**, such as if statements and for loops.
+
+Here, we'll aim to provide an overview of AutoGraph, as well as various
+restrictions and constraints you may discover.
+
+.. note::
+
+    For a more general overview of Catalyst, please see the
+    :doc:`quick start guide <quick_start>`.
+
+Using AutoGraph
+---------------
+
+AutoGraph currently requires TensorFlow as a dependency; in most cases it can
+be installed via
+
+.. code-block:: console
+
+    pip install tensorflow
+
+but please refer to the
+`TensorFlow documentation <https://www.tensorflow.org/install>`__
+for specific details on installing TensorFlow for your platform.
+
+Once TensorFlow is available, AutoGraph can be enabled by passing
+``autograph=True`` to the ``@qjit`` decorator:
+
+.. code-block:: python
+
+    dev = qml.device("lightning.qubit", wires=4)
+
+    @qjit(autograph=True)
+    @qml.qnode(dev)
+    def cost(weights, data):
+        qml.AngleEmbedding(data, wires=range(4))
+
+        for x in weights:
+
+            for j, p in enumerate(x):
+                if p > 0:
+                    qml.RX(p, wires=j)
+                elif p < 0:
+                    qml.RY(p, wires=j)
+
+            for j in range(4):
+                qml.CNOT(wires=[j, jnp.mod((j + 1), 4)])
+
+        return qml.expval(qml.PauliZ(0) + qml.PauliZ(3))
+
+>>> weights = jnp.linspace(-1, 1, 20).reshape([5, 4])
+>>> data = jnp.ones([4])
+>>> cost(weights, data)
+array(0.30455313)
+
+This would be equivalent to writing the following program, without using
+AutoGraph, but instead using :func:`~.cond` and :func:`~.for_loop`:
+
+.. code-block:: python
+
+    @qjit(autograph=False)
+    @qml.qnode(dev)
+    def cost(weights, data):
+        qml.AngleEmbedding(data, wires=range(4))
+
+        def layer_loop(i):
+            x = weights[i]
+            def wire_loop(j):
+
+                @cond(x[j] > 0)
+                def trainable_gate():
+                    qml.RX(x[j], wires=j)
+
+                @trainable_gate.else_if(x[j] < 0)
+                def trainable_gate():
+                    qml.RY(x[j], wires=j)
+
+                trainable_gate()
+
+            def cnot_loop(j):
+                qml.CNOT(wires=[j, jnp.mod((j + 1), 4)])
+
+            for_loop(0, 4, 1)(wire_loop)()
+            for_loop(0, 4, 1)(cnot_loop)()
+
+        for_loop(0, jnp.shape(weights)[0], 1)(layer_loop)()
+        return qml.expval(qml.PauliZ(0) + qml.PauliZ(3))
+
+>>> cost(weights, data)
+array(0.30455313)
+
+Currently, AutoGraph supports converting the following Python control
+flow statements:
+
+- ``if`` statements (including ``elif`` and ``else``)
+- ``for`` loops
+
+``while`` loops are currently not supported, alongside ``break`` and ``continue`` statements.
+
+In addition, ``lambda`` functions cannot currently be used with AutoGraph; please
+use standard function definitions instead.
```

Comment:
@dime10 @maliasadi this is the only change in this file not previously reviewed

Reply:
Looks good, thanks 👍